### ISSUE TITLE
opencl vs cuda migration

### ISSUE THREAD

**Issue by fpaboim** [geohot/tinygrad#484]
There are probably good reasons to migrate to cuda, e.g., better library support, debugging and such, but if the main reason is performance I doubt you will get considerable gains porting the current code. If you use higher level libraries then you could probably eventually get somethign similar to pytorch and tensorflow, as that's what they are using, but then you would give up one of the things which IMHO makes tinygrad most interesting. Also, the opencl code is slow because it doesn't use most of the parallelization techniques required for targeting GPUs like local memory, coalesced reads/writes, vectorization, complex instructions like FMA to save clocks, etc.. If you benckmark comparable cuda and opencl code the performance differences are not that great, certainly nothing close to two orders of magnitude which is what we see in some pytorch vs tinygrad benchmarks. Naive parallelization will only get you so far with GPGPU programming, or even targeting modern cpu like threadrippers or arm devices. I honestly thought of taking a jab at the reduction code but as it is today it's unreadable, some form of templating to make the kernel code more readable would help. 

That is all well and good but the mais reason to use opencl vs. cuda has only really been glossed over, which is it's support for multiple accelerators. If your endgame is making custom ai chips, like I think you've talked about in the past, opencl would actually be a good fit. If instead of targeting GPUs you target FPGAs you could probably get good performance, throughput all while being more power efficient which is important if you want to run the code on a portable device (like a dashcam thing) and tinygrad would be the perfect base for this. If you're really keen on making your own chip designs you could tune that nasty verilog code and use it as a starting point for an ASIC, you'd give up some flexiblity though. Still, it would require similar optimizations and the GPU code to work decently well. There is a bunch of stuff online, my own master's thesis was on GPGPU computing for finite element simulations and I included a tutorial on implementing such optimizations for the conjugate gradient method: https://github.com/fpaboim/gpufem/blob/master/Report/FEM%20with%20Sparse%20Matrices%20and%20OpenCL.docx

**Comment by geohot** [comment]
There's CUDA support in there now, it was a very small diff from OpenCL. I agree it's not why it's slower though, I already see big speedups with the optimizer by changing memory orderings and vectorizing. Going to push that further.
